Create EHR Certification ID features

nnobrigad opened the pull request:
API changes added new objects to support two new tables and Cert ID logic.

nnobrigad committed:
Added stubbed CertificationIds REST service

nnobrigad committed:
Added function getByIds() to retrieve multiple products by specifying multiple Ids.

nnobrigad committed:
Refactored product collection lookup.
Added more data in results.
Changed product id delimiter from semi-colon ( ; ) to pipe ( | )

nnobrigad committed:
Fixed getByIds query to lookup correct product_id field

nnobrigad committed:
Fixed missing parens in query

nnobrigad committed:
Removed unused ProductVersionManager

nnobrigad committed:
Merge branch 'development' of https://github.com/chpladmin/chpl-api into staging

nnobrigad committed:
Merge remote-tracking branch 'upstream/staging' into staging

nnobrigad committed:
Merge branch 'staging' of https://github.com/chpladmin/chpl-api into staging

nnobrigad committed:
Merge branch 'staging' of https://github.com/chpladmin/chpl-api into staging

nnobrigad committed:
Refactored EHR Certification ID controller code to lookup products using certified_product_details view instead of product table.

nnobrigad committed:
Add result size check to CertificationIdController Cert ID results.
Fixed format of returned Certification Id

nnobrigad committed:
Initial commit of EHR Certification ID code and objects.

nnobrigad committed:
Completed first pass 2014 EHR Certification ID product collection validation

nnobrigad committed:
Refactored CertID validation to separate package/class

nnobrigad committed:
Refactored EHR Certification ID collection validation and lookup
Added validator for 2015 and 2014/2015 hybrid years
Added Year field to result products and certification id data
Added temporary ID generation code

nnobrigad committed:
New EHR Cert IDs now persist
Replaced data object yearId fields with year text fields

nnobrigad committed:
Added create parameter for getCertificationId only create new IDs when requested

andlar commented:
# Comments on pull requests:

## Generally:
- For every pull-request / commit, please add something to the `RELEASE_NOTES.md` file to indicate what changed. (For the data-model it'd likely be something like: "Added new tables to support CMS Certification ID"). Generally we put the new items in a "Next version" section, with a Date that's TBD, and without indicating what the new version number will actually be, as that is determined at deployment time.

## api:
- chpl-service/src/main/java/gov/healthit/chpl/dao/impl/CertificationIdDAOImpl.java
  - Since you have to hardcode the user ID because these calls may be run by unauthenticated users, we'd like you to use -4 instead of -1. We are using the user with ID -1 for other purposes in our preloaded data, -2 is our default admin user, -3 is used for another unauthenticated call (creating API keys), and -4 is next on the list.
  - The delete method does not appear to set the deleted boolean to true.
- chpl-service/src/main/java/gov/healthit/chpl/manager/impl/CertificationIdManagerImpl.java
  - It looks like the delete call would insert two "deleted" statements into the activity table.
  - Is or will anything be calling the update or delete methods in the CertificationIdManagerImpl? It's fine to leave them in there if that's the plan but I didn't see them being used currently.
- chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/impl/ProductManagerImpl.java
  - I don't see the getByIds method being called anywhere. If that's true, could this code (and the interface method and the DAO method it references) all be removed?
- chpl-service/src/main/java/gov/healthit/chpl/web/controller/CertificationIdController.java
  - Could you make the main rest endpoint camelCase? So "certificationIds" instead of "certificationids" to be consistent with our other REST endpoints.
  - You may want to give the API endpoint /getCertificationId a different URL. The request is of type GET so putting "get" in the endpoint URL ends up being repetitive. We typically use the endpoint "/" for our basic "return everything" method. For example calling an HTTP GET on /developers would return all the developers. Parameters in the URL or query string beyond that act as filters to those results. So you might want /certificationIds/?products=1|2|3|4 to return the certification IDs for those four products. This would make the public-facing API endpoint consistent with our existing ones. GET /certificationIds/{certificationId} is perfect.
- We use the Swagger library to display our API documentation. I see that you have the controller methods annotated with API notes which is great, but there is one more non-obvious file that has to be changed to get these docs to appear. You'll need to add the base URL to gov.healthit.chpl.SwaggerConfig. If you can run the CHPL website locally, you should see your API calls listed with the others under the "Resources" link at the top of the screen.

nnobrigad committed:
Updated CPOE and Direct Project criteria listings

nnobrigad committed:
Fixed order of Transitions of Care Criteria checks from most criteria to least criteria required
Changed CQM Criteria checks to be "all required"
Changed Privacy and Security Criteria checks to be "all required"

nnobrigad committed:
Changed modified user id from -1 to -4 for Certification ID DB insert operations.

nnobrigad committed:
Fixed missing semicolon.

nnobrigad committed:
Removed extraneous defunct methods from ProductManager and ProductDAO

nnobrigad committed:
Removed extraneous methods delete and update from CertificationID Manager as this data should not be mutable once created

nnobrigad committed:
Changed Certification ID REST endpoint to camel-case ("certificationIds")
Changed GET name from "/getCertificationId" to "/"

nnobrigad committed:
Changed Certification ID endpoint to "certification_ids" to match the convention of other service endpoint names.

nnobrigad committed:
Removed delete and update methods from Certification ID DAO and Entities as Certification IDs should not be mutable

nnobrigad committed:
Updated release notes for adding EHR Certification Id features

nnobrigad commented:
- Changed modified user from -1 to -4
- Removed unused delete() and update() methods for the ehr_certification_id table DAOs and others classes as these records should be immutable once created.
- Removed unused methods from ProductManager.
- Changed the endpoint naming to match other service endpoints using an underscore ("certification_ids")
- Changed the endpoint to root "/" instead of "getCertificationId"
- Added service to Swagger config.
- Fixed a few bugs in product collection validators and controller.

nnobrigad committed:
Merge branch 'development' of https://github.com/andlar/chpl-api into staging
